include/envoy/thread/thread.h

```diff
@@ -2,40 +2,149 @@
 
 #include "envoy/common/pure.h"
 
+#include "absl/base/thread_annotations.h"
+
 namespace Envoy {
 namespace Thread {
 
 /**
- * Like the C++11 "basic lockable concept" but a concrete interface vs. a template.
+ * Like the C++11 "basic lockable concept" but a pure virtual interface vs. a template, and
+ * with thread annotations.
  */
-class BasicLockable {
+class LOCKABLE BasicLockable {
 public:
   virtual ~BasicLockable() {}
 
-  virtual void lock() PURE;
-  virtual bool try_lock() PURE;
-  virtual void unlock() PURE;
+  virtual void lock() EXCLUSIVE_LOCK_FUNCTION() PURE;
+  virtual bool tryLock() EXCLUSIVE_TRYLOCK_FUNCTION(true) PURE;
+  virtual void unlock() UNLOCK_FUNCTION() PURE;
 };
 
 /**
  * A lock guard that deals with an optional lock.
  */
-template <class T> class OptionalLockGuard {
+class SCOPED_LOCKABLE OptionalLockGuard {
 public:
-  OptionalLockGuard(T* lock) : lock_(lock) {
-    if (lock) {
-      lock->lock();
+  /**
+   * Establishes a scoped mutex-lock. If non-null, the mutex is locked upon construction.
+   *
+   * @param lock the mutex.
+   */
+  OptionalLockGuard(BasicLockable* lock) EXCLUSIVE_LOCK_FUNCTION(lock) : lock_(lock) {
+    if (lock_) {
+      lock_->lock();
     }
   }
 
-  ~OptionalLockGuard() {
+  /**
+   * Destruction of the OptionalLockGuard unlocks the lock, if it is non-null.
+   */
+  ~OptionalLockGuard() UNLOCK_FUNCTION() {
     if (lock_) {
       lock_->unlock();
     }
   }
 
 private:
-  T* lock_;
+  BasicLockable* lock_;
+};
+
+/**
+ * A lock guard that deals with a lock that is not locked on construction, although
+ * it is unlocked on destruction, if necessary.
+ */
+class DeferredLockGuard {
+public:
+  /**
+   * Establishes a scoped mutex-lock; the mutex is not locked upon construction.
+   *
+   * @param lock the mutex.
+   */
+  DeferredLockGuard(BasicLockable& lock) : lock_(lock), is_locked_(false) {}
+
+  /**
+   * Destruction of the DeferredLockGuard unlocks the lock, if it was locked.
+   */
+  ~DeferredLockGuard() {
+    if (is_locked_) {
+      lock_.unlock();
+    }
+  }
+
+  // Attempts to lock the mutex, if present. Returns false if no lock was taken.
+  bool tryLock() {
```

Comment:
Does this class not need annotations?

Reply:
Actually it's worse than that.  I can't figure out how to annotate this class, as I haven't found a way to express "the destructor should unlock the lock if the tryLock succeeded" in a way that actually works.

I don't totally understand, TBH, how it's able to handle ReleasableLockGuard and not this, as that's also a confusing concept in terms of static analysis.  The reason that things compile like this is that the current PR state is that in hot_restart_impl.h, there's no GUARDED_BY variables for init_lock.  Actually I'm not sure which variables should be guarded by init_lock, if any.  It seems that most of those variable that are accessed with the lock are also accessed without it.

I wonder if it's better to remove this class from include/envoy/thread/thread.h and just put it where it's used, in hot_restart_impl.cc, as it's sitting here amongst some superficially similar but better-annotated LockGuard variants.  There we can ensure the lock gets released in the presence of an exception, and ignore thread annotations for that particular mutex.

Reply:
I decided to bail on DeferredLockGuard.  I was able to get it to work in https://github.com/envoyproxy/envoy/pull/3321/commits/c4247542a75d77c0db250766984c575ce3ce3b06 but I had to disable lock analysis within the body of tryLock(), which was annoying.

I now provide TryLockGuard, which runs a tryLock() on construction, and provides an isLocked() predicate.  This required changing the call-site slightly, but it still seems reasonable to me. The advantage is that thread annotations work without having to disable them.

WDYT?